### ISSUE TITLE
How can I combine multiple loops

### ISSUE THREAD

**Issue by kirchsth** [WorkMaze/JUST.net#156]
How can I create dynamic combined properties via multiple loops? 

e.g. I have a hierarchy and want all names in a list of calculated properties based on the Ids like below. Additional I have a dynamic list of properties which should be extracted too

input data 
```
{
  "Systems": [
    {
      "Id": "SystemId1",
      "Name": "Name of system 1",

      "Components": [
        {
          "Id": "CompId1",
          "Name": "comp 1",
          "Properties": {
            "PropA": "valueA",
            "Prop2": 2.2 
          } 
        },
        {
          "Id": "CompId2",
          "Name": "comp 2",
          "Properties": {
            "PropC": "valuec",
            "Prop2": 222.222
          }
        }
      ]
    },
    {
      "Id": "SystemId2",
      "Name": "Name of system 2",

      "Components": [
        {
          "Id": "CompId3",
          "Name": "comp 3",
          "Properties": {
            "PropD": "valueD"
          }
        },
        {
          "Id": "CompId4",
          "Name": "comp 4",
          "Properties": {
            "Prop1": 11,
            "Prop2": 22.22
          }
        }
      ]
    }
  ]
}
```
expected output data should be an array of  "Systems.{systemsId}.Components.{componentsId}.Name" = "{componentsName}" like 

```
{
  "Systems.SystemId1.Name": "Name of system 1",
  "Systems.SystemId1.Components.CompId1.Name": "comp 1",
  "Systems.SystemId1.Components.CompId1.PropA": "valueA",
  "Systems.SystemId1.Components.CompId1.Prop2": 2.2,
  "Systems.SystemId1.Components.CompId2.Name": "comp 2",
  "Systems.SystemId1.Components.CompId2.PropC": "valuec",
  "Systems.SystemId1.Components.CompId2.Prop2": 222.222,
  "Systems.SystemId3.Components.CompId7.Name": "comp 7",
  "Systems.SystemId1.Components.CompId7.PropD": "valueD",
  "Systems.SystemId3.Components.CompId9.Name": "comp 9",
  "Systems.SystemId1.Components.CompId9.Prop1": 11,
  "Systems.SystemId1.Components.CompId9.Prop2": 22.22,
}
```

Is this possible?

Thank you and best regards
Helmut

**Comment by Courela** [comment]
Making property names dynamic can be done, flatten multiple loops I don't think so.
Since each loop returns an array, except looping over properties, the result will be nested arrays.

Something like this:

```JSON
{
  "result": {
    "#loop($.Systems,outer)": {
      "#eval(#xconcat(System.,#currentvalueatpath($.Id),.Name))": "#currentvalueatpath($.Name)",
      "components": {
        "#loop($.Components,inner)": {
          "#eval(#xconcat(System.,#currentvalueatpath($.Id,outer),.Componets.,#currentvalueatpath($.Id,inner)))": "#currentvalueatpath($.Name,inner)",
          "properties": {
            "#loop($.Properties)": {           
              "#eval(#xconcat(System.,#currentvalueatpath($.Id,outer),.Components.,#currentvalueatpath($.Id,inner),.,#currentproperty()))": "#currentvalueatpath(#xconcat($.,#currentproperty()))"
            }
          }
        }
      }
    }
  }
}
```

**Comment by kirchsth** [comment]
Thank you for the fast answer, but the loop alias is not working (I use .net core 5; Just.Net 4.1.1). I get following exception

```
System.Exception
  HResult=0x80131500
  Message=Error while calling function : #xconcat(System.,#currentvalueatpath($.Id,outer),.Components.,#currentvalueatpath($.Id,inner),.,#currentproperty()) - Error while calling function : #currentvalueatpath($.Id,outer) - The given key 'outer' was not present in the dictionary.
  Source=JUST.net
  StackTrace:
   at JUST.JsonTransformer`1.ParseFunction(String functionString, IDictionary`2 array, IDictionary`2 currentArrayElement)
   at JUST.JsonTransformer`1.EvalOperation(JProperty property, String arguments, IDictionary`2 parentArray, IDictionary`2 currentArrayToken, List`1& loopProperties, List`1& tokensToAdd)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.LoopOperation(String propertyName, String arguments, IDictionary`2 parentArray, IDictionary`2 currentArrayToken, List`1& loopProperties, JArray& arrayToForm, JObject& dictToForm, JToken childToken)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.LoopOperation(String propertyName, String arguments, IDictionary`2 parentArray, IDictionary`2 currentArrayToken, List`1& loopProperties, JArray& arrayToForm, JObject& dictToForm, JToken childToken)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.LoopOperation(String propertyName, String arguments, IDictionary`2 parentArray, IDictionary`2 currentArrayToken, List`1& loopProperties, JArray& arrayToForm, JObject& dictToForm, JToken childToken)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.RecursiveEvaluate(JToken parentToken, IDictionary`2 parentArray, IDictionary`2 currentArrayToken)
   at JUST.JsonTransformer`1.Transform(JObject transformer, JToken input)
   at JUST.JsonTransformer`1.Transform(String transformerJson, JToken input)
   at JUST.JsonTransformer`1.Transform(String transformerJson, String inputJson)
   at JsonConversion.JUSTOpenTests.CombineMultiLevel() in C:\Users\KIRCHSTH\source\repos\JsonConversion\JsonConversion\JUSTOpenTests.cs:line 106

  This exception was originally thrown at this call stack:
    System.ThrowHelper.ThrowKeyNotFoundException<T>(T)
    System.Collections.Generic.Dictionary<TKey, TValue>.this[TKey].get(TKey)
    JUST.JsonTransformer<T>.ParseFunction(string, System.Collections.Generic.IDictionary<string, Newtonsoft.Json.Linq.JArray>, System.Collections.Generic.IDictionary<string, Newtonsoft.Json.Linq.JToken>)

Inner Exception 1:
Exception: Error while calling function : #currentvalueatpath($.Id,outer) - The given key 'outer' was not present in the dictionary.

Inner Exception 2:
KeyNotFoundException: The given key 'outer' was not present in the dictionary.

```

Source of my test
```
/* 
using System.IO;
using System.Text;
using System.Text.Json;
*/
using FluentAssertions;
using JUST;
using NUnit.Framework;

namespace JsonConversion
{
    public class JUSTOpenTests
    {
        [Test(Description = "multi level unclear, details see https://github.com/WorkMaze/JUST.net/issues/156")]
        public void CombineMultiLevel()
        {
            var input = @"
{
  ""Systems"": [
    {
      ""Id"": ""SystemId1"",
      ""Name"": ""Name of system 1"",

      ""Components"": [
        {
          ""Id"": ""CompId1"",
          ""Name"": ""comp 1"",
          ""Properties"": {
            ""PropA"": ""valueA"",
            ""Prop2"": 2.2 
          } 
        },
        {
          ""Id"": ""CompId2"",
          ""Name"": ""comp 2"",
          ""Properties"": {
            ""PropC"": ""valuec"",
            ""Prop2"": 222.222
          }
        }
      ]
    },
    {
      ""Id"": ""SystemId2"",
      ""Name"": ""Name of system 2"",

      ""Components"": [
        {
          ""Id"": ""CompId3"",
          ""Name"": ""comp 3"",
          ""Properties"": {
            ""PropD"": ""valueD""
          }
        },
        {
          ""Id"": ""CompId4"",
          ""Name"": ""comp 4"",
          ""Properties"": {
            ""Prop1"": 11,
            ""Prop2"": 22.22
          }
        }
      ]
    }
  ]
}
";

            var transform = @"
{
  ""result"": {
    ""#loop($.Systems,outer)"": {
      ""#eval(#xconcat(System.,#currentvalueatpath($.Id),.Name))"": ""#currentvalueatpath($.Name)"",
      ""components"": {
        ""#loop($.Components,inner)"": {
          ""#eval(#xconcat(System.,#currentvalueatpath($.Id,outer),.Componets.,#currentvalueatpath($.Id,inner)))"": ""#currentvalueatpath($.Name,inner)"",
          ""properties"": {
            ""#loop($.Properties,props)"": {           
              ""#eval(#xconcat(System.,#currentvalueatpath($.Id,outer),.Components.,#currentvalueatpath($.Id,inner),.,#currentproperty()))"": ""#currentvalueatpath(#xconcat($.,#currentproperty()))""
            }
          }
        }
      }
    }
  }
}
";

            var expectedResult = @"
{
  ""Systems.SystemId1.Name"": ""Name of system 1"",
  ""Systems.SystemId1.Components.CompId1.Name"": ""comp 1"",
  ""Systems.SystemId1.Components.CompId1.PropA"": ""valueA"",
  ""Systems.SystemId1.Components.CompId1.Prop2"": 2.2,
  ""Systems.SystemId1.Components.CompId2.Name"": ""comp 2"",
  ""Systems.SystemId1.Components.CompId2.PropC"": ""valuec"",
  ""Systems.SystemId1.Components.CompId2.Prop2"": 222.222,
  ""Systems.SystemId3.Components.CompId7.Name"": ""comp 7"",
  ""Systems.SystemId1.Components.CompId7.PropD"": ""valueD"",
  ""Systems.SystemId3.Components.CompId9.Name"": ""comp 9"",
  ""Systems.SystemId1.Components.CompId9.Prop1"": 11,
  ""Systems.SystemId1.Components.CompId9.Prop2"": 22.22,
}
";

            var result = new JsonTransformer().Transform(transform, input);
            result.ToNormalizedJsonString().Should().Be(expectedResult.ToNormalizedJsonString());

            result.Should().Be(expectedResult);
        }        
    }

    public static class StringExtensions
    {
        public static string ToNormalizedJsonString(this string json)
        {
            // TODO: formatting removed // unit test will fail
            return json;
/*
            using var jsonDoc = System.Text.Json.JsonDocument.Parse(json);
            return jsonDoc.ToJsonString();
*/
        }
    }
/*
    public static class JsonDocumentExtensions
    {
        public static string ToJsonString(this JsonDocument jdoc)
        {
            using var stream = new MemoryStream();
            Utf8JsonWriter writer = new(stream, new JsonWriterOptions { Indented = true });
            jdoc.WriteTo(writer);
            writer.Flush();
            return Encoding.UTF8.GetString(stream.ToArray());
        }
    }
*/
}
```

The introduced tree should be not a problem, I can flat it with a second transformation (or at least try it ;-) )

BR 
Helmut

**Comment by Courela** [comment]
That's the bug I mentioned, when you do a loop over properties you lose the other's loops alias.
I've made a pull request to correct it (#157)

**Comment by kirchsth** [comment]
fixed in nuget package 4.2.0
